feat!: (fix + chore) streamline nSegments usage

asalzburger opened the pull request:
The generation of display vertices for segments was buggy and very unstable, I have put this onto more solid grounds, which helps displaying (not only) `AnnulusBounds` correctly.

ITk Petal Before:

![Screenshot 2024-07-18 at 17 01 59](https://github.com/user-attachments/assets/930b93f2-8df7-4e80-a44b-3ec7a20a84ac)

ITk Petal After:

![Screenshot 2024-07-19 at 12 07 32](https://github.com/user-attachments/assets/bc4edd4b-cc22-4b30-bad8-6032d42854da)

When checking the code, I see that there was a pretty big inconsistency in how the nSegments have been handled, so I re-worked that:

* `nSegments` (full 2 * PI segments) -> changed and renamed to `quaterSegments`

This guarantees that the number of segments for a full circle is a multiplier of 4 and hence the phi extrema points at `(-pi,-0.5*pi,0,0.5*pi,pi)` are consistently added. This will lead to a correct `x` and `y` estimation for the `Extent` of the `Polyhedron`.

This also guarantees that e.g. overlapping/attaching surfaces with the same number of segments will have vertices at the same positions, which will be better for the displaying.

* inconsitent use of `unsigned int` (SurfaceBounds) and `std::size_t` -> changed to `unsigned int everywhere

* Introduces tests for the vertex generation (which were missing)

paulgessinger reviewed (changes requested):
Can factorize the fix to the vertex drawing from the other changes (general cleanup)?

paulgessinger commented:
I was wondering generally: wouldn't we be able to get rid of manual passing of segments if we just define a maximum arc length $r\times\phi$ in $mm$ that a single segment is allowed to be? That way, the target bounds could decide on their own how many segments to produce, because they can just calculate the arc length?

paulgessinger commented on Core/include/Acts/Detector/LayerStructureBuilder.hpp:
```diff
@@ -91,8 +91,9 @@ class LayerStructureBuilder : public IInternalStructureBuilder {
-    unsigned int nSegments = 1u;
+    /// Polyhedron approximations: number of segments to be used
+    /// to approximate a quater of a circle
+    unsigned int quaterSegments = 1u;
```
```suggestion
    unsigned int quarterSegments = 1u;
```

asalzburger replied:
uuuuupsi.

paulgessinger commented on Core/include/Acts/Visualization/ViewConfig.hpp:
```diff
@@ -37,8 +37,8 @@ struct ViewConfig {
-  /// The number of segments to approximate full 2pi
-  unsigned int nSegments = 72;
+  /// The number of segments to approximate a quarter of the circle
+  unsigned int quaterSegments = 72;
```
```suggestion
  unsigned int quarterSegments = 72;
```

paulgessinger commented on Core/src/Surfaces/detail/VerticesHelper.cpp:
Three's not really a need for a `detail` folder in the sources, since they don't get installed.

asalzburger replied:
I know - it's just that we had the other helps in `detail` so I at least wanted it consistent.

paulgessinger replied:
ok

asalzburger commented:
> Can you avoid the physmon output changes? This should not affect the reconstruction at all should it?

I suppose that we get a slightly different bin association for one or another surface now ... hence the fluctuation in the output.

paulgessinger commented:
@asalzburger Is this because we use it for the auto-binning in the Gen1 layer building?

asalzburger commented:
> I was wondering generally: wouldn't we be able to get rid of manual passing of segments if we just define a maximum arc length r×ϕ in mm that a single segment is allowed to be? That way, the target bounds could decide on their own how many segments to produce, because they can just calculate the arc length?

That's a possibility, however:
The polyhedron is used mainly in two different context situations:
-  (a) displaying
-  (b) space/extent estimation

For latter (in many cases) a choice of `quarterSegments =1` is sufficient, as it will generate now the phi values that are need to estimate the maximum extent in x/y.  That's the reason why I have changed to `quarterSegments` rather than full 2*pi.

For displaying, however, this newly implemented schema - where the number of segments is given per quarter (and drives the vertex generation) will guarantee that attaching surfaces (even with different phi sector openings) will have vertices at the same points in space and hence will "look attaching".

That's why I changed to this schema.

asalzburger commented:
> @asalzburger Is this because we use it for the auto-binning in the Gen1 layer building?

That is my assumption, I will try to dig down a bit on this - I was surprised as well ....

asalzburger commented:
Hi @paulgessinger @noemina - I think this is ready, the performance changes are also gone

paulgessinger commented:
@asalzburger awesome! Does this have time so I can look through it tomorrow?

andiwand commented on Core/include/Acts/Surfaces/AnnulusBounds.hpp:
```diff
@@ -133,19 +133,20 @@ class AnnulusBounds : public DiscBounds {
 
   /// This method returns the xy coordinates of the four corners of the
-  /// bounds in module coordinates (in x/y)
+  /// bounds in module coordinates (in x/y), and if quarterSegments is bigger of
```
```suggestion
  /// bounds in module coordinates (in x/y), and if quarterSegments is bigger or
```

andiwand commented on Core/include/Acts/Surfaces/AnnulusBounds.hpp:
```diff
@@ -133,19 +133,20 @@ class AnnulusBounds : public DiscBounds {
-  /// bounds in module coordinates (in x/y)
+  /// bounds in module coordinates (in x/y), and if quarterSegments is bigger of
+  /// equal to 0, the curved part of the segment is included and approximated
+  /// by the correspondingnumber of segments.
```
```suggestion
  /// by the corresponding number of segments.
```

andiwand commented on Core/include/Acts/Surfaces/ConvexPolygonBounds.hpp:
```diff
@@ -112,13 +112,13 @@ class ConvexPolygonBounds : public ConvexPolygonBoundsBase {
   ///
   /// @return vector for vertices in 2D
-  std::vector<Vector2> vertices(unsigned int lseg = 1) const final;
+  std::vector<Vector2> vertices(unsigned int ignoredSegments = 0u) const final;
```
Interface wise this integer has a different meaning depending on the surface right? I wonder if that might be surprising if you call this method from the base class. Potentially all the different params could be bundled up in a struct and passed as options?

asalzburger replied:
For surfaces without curves it's actually ignored - which is why it is renamed.

andiwand commented on Core/include/Acts/Surfaces/CylinderSurface.hpp:
```diff
@@ -214,14 +214,20 @@ class CylinderSurface : public RegularSurface {
-  /// the full 2*M_PI coverange, if lseg is set to 1 only the extrema
-  /// are given
+  /// @param quarterSegments The number of segments to approximate a quarter of the
+  /// full circle; it it's chosen to be 1, only the extrema points (-pi, -0.5pi,
```
```suggestion
  /// full circle; it's chosen to be 1, only the extrema points (-pi, -0.5pi,
```

andiwand commented on Core/include/Acts/Surfaces/detail/VerticesHelper.hpp:
```diff
@@ -22,51 +22,53 @@ namespace Acts::detail::VerticesHelper {
 
 /// Helper method to create a regular 2 or 3 D segment
-///  between two phi values
+///  between two phi values with a given number of segments
```
```suggestion
/// between two phi values with a given number of segments
```

andiwand commented on Core/include/Acts/Surfaces/detail/VerticesHelper.hpp:
```diff
@@ -22,51 +22,53 @@ namespace Acts::detail::VerticesHelper {
-/// @param addon The additional segments to be built
+/// @param rXY The radius description if first +/= second: ellipse
+/// @param phiMin the minimum phi value
+/// @param phiMax The second phi value
```
```suggestion
/// @param phiMax the second phi value
```

andiwand commented on Core/src/Surfaces/ConeSurface.cpp:
```diff
@@ -208,64 +207,54 @@ Acts::Polyhedron Acts::ConeSurface::polyhedronRepresentation(
-  double avgPhi = bounds().get(ConeBounds::eAveragePhi);
+  ActsScalar hPhiSec = bounds().get(ConeBounds::eHalfPhiSector);
+  ActsScalar avgPhi = bounds().get(ConeBounds::eAveragePhi);
   bool fullCone = (hPhiSec == M_PI);
+  std::vector<ActsScalar> refPhi = {};
+  if (not fullCone) {
```
```suggestion
  std::vector<ActsScalar> refPhi = {};
  if (bool fullCone = (hPhiSec == M_PI); !fullCone) {
```

andiwand commented on Core/src/Surfaces/CylinderBounds.cpp:
```diff
@@ -104,45 +104,41 @@ std::ostream& Acts::CylinderBounds::toStream(std::ostream& sl) const {
-                                    avgPhi - halfPhi, avgPhi + halfPhi,
-                                    {static_cast<ActsScalar>(avgPhi)});
+  std::vector<ActsScalar> phiRef = {};
+  if (fullCylinder) {
```
```suggestion
  if (bool fullCylinder = coversFullAzimuth(); fullCylinder) {
```

andiwand commented on Core/src/Surfaces/CylinderBounds.cpp:
```diff
@@ -104,45 +104,41 @@ std::ostream& Acts::CylinderBounds::toStream(std::ostream& sl) const {
   double avgPhi = get(eAveragePhi);
   double halfPhi = get(eHalfPhiSector);
 
   bool fullCylinder = coversFullAzimuth();
```
```suggestion
```

andiwand commented on Core/src/Surfaces/PlaneSurface.cpp:
```diff
@@ -94,13 +94,14 @@ const Acts::SurfaceBounds& Acts::PlaneSurface::bounds() const {
+  unsigned int lseg = quarterSegments;
   // If you have bounds you can create a polyhedron representation
   if (m_bounds) {
     auto vertices2D = m_bounds->vertices(lseg);
```
```suggestion
    auto vertices2D = m_bounds->vertices(quarterSegments);
```

andiwand commented on Core/src/Surfaces/PlaneSurface.cpp:
```diff
@@ -94,13 +94,14 @@ const Acts::SurfaceBounds& Acts::PlaneSurface::bounds() const {
   std::vector<Polyhedron::FaceType> triangularMesh;
   bool exactPolyhedron = true;
 
+  unsigned int lseg = quarterSegments;
```
```suggestion
```

andiwand commented on Examples/Python/src/Obj.cpp:
```diff
@@ -34,16 +34,16 @@ void addObj(Context& ctx) {
     /// @param viewContext is the geometry context
     /// @param viewRgb is the color of the surfaces
-    /// @param viewSegements is the number of segments to approximate a full circle
+    /// @param viewsegments is the number of segments to approximate a quarter of a circle
```
```suggestion
    /// @param viewSegments is the number of segments to approximate a quarter of a circle
```

andiwand commented on Examples/Python/src/Obj.cpp:
```diff
@@ -34,16 +34,16 @@ void addObj(Context& ctx) {
             [](const std::vector<std::shared_ptr<Surface>>& surfaces,
                const GeometryContext& viewContext,
-               const std::array<int, 3>& viewRgb, unsigned int viewSegements,
+               const std::array<int, 3>& viewRgb, unsigned int viewsegments,
```
```suggestion
               const std::array<int, 3>& viewRgb, unsigned int viewSegments,
```

andiwand commented on Examples/Python/src/Obj.cpp:
```diff
@@ -34,16 +34,16 @@ void addObj(Context& ctx) {
                const std::string& fileName) {
               Acts::ViewConfig sConfig = Acts::ViewConfig{viewRgb};
-              sConfig.nSegments = viewSegements;
+              sConfig.quarterSegments = viewsegments;
```
```suggestion
              sConfig.quarterSegments = viewSegments;
```

andiwand commented on Tests/UnitTests/Fatras/Digitization/SegmentizerTests.cpp:
```diff
@@ -217,9 +217,9 @@ BOOST_DATA_TEST_CASE(
 
     /// Run the Segmentizer
-    auto cSegement = cl.segments(geoCtx, *surface, segmentation, {start, end});
+    auto csegment = cl.segments(geoCtx, *surface, segmentation, {start, end});
```
```suggestion
    auto cSegments = cl.segments(geoCtx, *surface, segmentation, {start, end});
```

andiwand commented on Tests/UnitTests/Fatras/Digitization/SegmentizerTests.cpp:
```diff
@@ -217,9 +217,9 @@ BOOST_DATA_TEST_CASE(
+    auto csegment = cl.segments(geoCtx, *surface, segmentation, {start, end});
 
-    for (const auto& cs : cSegement) {
+    for (const auto& cs : csegment) {
```
```suggestion
    for (const auto& cs : cSegments) {
```

andiwand reviewed (commented):
had a quick look - mostly nitpicks from my side

paulgessinger commented on Core/include/Acts/Surfaces/detail/VerticesHelper.hpp:
```diff
@@ -22,51 +22,53 @@ namespace Acts::detail::VerticesHelper {
+  std::vector<vertex_t> vertices;
+  std::vector<ActsScalar> phis =
+      phiSegments(phiMin, phiMax, phiRefs, quarterSegments);
+  for (auto phi : phis) {
```
```suggestion
  for (ActsScalar phi : phis) {
```

paulgessinger commented on Core/include/Acts/Surfaces/Surface.hpp:
```diff
@@ -420,20 +420,21 @@ class Surface : public virtual GeometryObject,
+  /// which represents a quarter of the full circle
+  ///
+  /// @note In order to symmetrize the code between sectoral and closed cylinders
+  /// in case of closed cylinders, both (-pi, pi) are given as separate vertices
```
I guess this can be kind of glitchy in visualization, but probably not a big deal.

asalzburger replied:
I checked on the obj, no effect - and it implies the code quite a bit as I do not have to branch between closed and open cylinders.

paulgessinger commented on Core/src/Surfaces/detail/VerticesHelper.cpp:
```diff
@@ -13,44 +13,63 @@
+  }
+
+  // First check that no reference phi is outside the range
+  for (const auto& phiRef : phiRefs) {
```
```suggestion
  for (ActsScalar phiRef : phiRefs) {
```

paulgessinger commented on Core/src/Surfaces/detail/VerticesHelper.cpp:
```diff
@@ -13,44 +13,63 @@
-          });
-      if (match == phiSegments.end()) {
+  // Bail out if minSegmens is smaller 4 or not a multiple of 4
+  if (quarterSegments == 0u) {
```
The comment does not line up with the actual check. What's correct?

paulgessinger commented on Core/src/Surfaces/PlaneSurface.cpp:
```diff
@@ -129,7 +129,7 @@ Acts::Polyhedron Acts::PlaneSurface::polyhedronRepresentation(
       // but that creates too  many unneccesarry faces, when only two
       // are needed to describe the mesh, @todo investigate merging flag
-      auto facesMesh = detail::FacesHelper::cylindricalFaceMesh(vertices, true);
+      auto facesMesh = detail::FacesHelper::cylindricalFaceMesh(vertices);
       faces = facesMesh.first;
       triangularMesh = facesMesh.second;
```
```suggestion
      auto [faces, triangularMesh] = detail::FacesHelper::cylindricalFaceMesh(vertices);
```
There's a similar case a few lines up that's not part of the changeset.

andiwand commented:
there is a small build issue @paulgessinger 

```
[232/1487] Building CXX object Core/CMakeFiles/ActsCore.dir/src/Surfaces/PlaneSurface.cpp.o
FAILED: Core/CMakeFiles/ActsCore.dir/src/Surfaces/PlaneSurface.cpp.o 
ccache /usr/bin/c++ -DACTS_ENABLE_LOG_FAILURE_THRESHOLD -DActsCore_EXPORTS -I/__w/acts/acts/Core/include -I/__w/acts/acts/build/Core -isystem /__w/acts/acts/cmake/assert_include -isystem /usr/include/eigen3 -Wall -Wextra -Wpedantic -Wshadow -Wzero-as-null-pointer-constant -Wold-style-cast -Werror -O3 -DNDEBUG -std=c++20 -fPIC -MD -MT Core/CMakeFiles/ActsCore.dir/src/Surfaces/PlaneSurface.cpp.o -MF Core/CMakeFiles/ActsCore.dir/src/Surfaces/PlaneSurface.cpp.o.d -o Core/CMakeFiles/ActsCore.dir/src/Surfaces/PlaneSurface.cpp.o -c /__w/acts/acts/Core/src/Surfaces/PlaneSurface.cpp
/__w/acts/acts/Core/src/Surfaces/PlaneSurface.cpp: In member function 'virtual Acts::Polyhedron Acts::PlaneSurface::polyhedronRepresentation(const Acts::GeometryContext&, unsigned int) const':
/__w/acts/acts/Core/src/Surfaces/PlaneSurface.cpp:119:13: error: declaration of 'faces' shadows a previous local [-Werror=shadow]
  119 |       auto [faces, triangularMesh] =
      |             ^~~~~
/__w/acts/acts/Core/src/Surfaces/PlaneSurface.cpp:94:37: note: shadowed declaration is here
   94 |   std::vector<Polyhedron::FaceType> faces;
      |                                     ^~~~~
/__w/acts/acts/Core/src/Surfaces/PlaneSurface.cpp:119:20: error: declaration of 'triangularMesh' shadows a previous local [-Werror=shadow]
  119 |       auto [faces, triangularMesh] =
      |                    ^~~~~~~~~~~~~~
/__w/acts/acts/Core/src/Surfaces/PlaneSurface.cpp:95:37: note: shadowed declaration is here
   95 |   std::vector<Polyhedron::FaceType> triangularMesh;
      |                                     ^~~~~~~~~~~~~~
/__w/acts/acts/Core/src/Surfaces/PlaneSurface.cpp:126:13: error: declaration of 'faces' shadows a previous local [-Werror=shadow]
  126 |       auto [faces, triangularMesh] =
      |             ^~~~~
/__w/acts/acts/Core/src/Surfaces/PlaneSurface.cpp:94:37: note: shadowed declaration is here
   94 |   std::vector<Polyhedron::FaceType> faces;
      |                                     ^~~~~
/__w/acts/acts/Core/src/Surfaces/PlaneSurface.cpp:126:20: error: declaration of 'triangularMesh' shadows a previous local [-Werror=shadow]
  126 |       auto [faces, triangularMesh] =
      |                    ^~~~~~~~~~~~~~
/__w/acts/acts/Core/src/Surfaces/PlaneSurface.cpp:95:37: note: shadowed declaration is here
   95 |   std::vector<Polyhedron::FaceType> triangularMesh;
      |                                     ^~~~~~~~~~~~~~
cc1plus: all warnings being treated as errors
```

paulgessinger commented:
@andiwand meh, I thought that was because of the merge conflict. I'll fix it.

paulgessinger commented:
@andiwand Compile fixes should be in now.

andiwand commented:
@paulgessinger can you also approve?
![image](https://github.com/user-attachments/assets/61add4b5-7646-4d63-a833-cfd4749d190b)

asalzburger reviewed (commented):
Good to go from my side.

acts-project-service commented:
## :red_circle: Athena integration test results [f5a34f31238f541d27b0f1238948818389e3455b]
### Build job with this PR failed!
Please investigate the [build job](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/44286101) for the [pipeline](https://gitlab.cern.ch/acts/acts-athena-ci/-/pipelines/8253781)!
